core/message_thread_manager.py

```diff
@@ -0,0 +1,176 @@
+        self.conn.commit()
+
+    def get_message(self, thread_id: int, message_index: int) -> Optional[Dict[str, Any]]:
+        self.cursor.execute("SELECT messages FROM ThreadMessages WHERE thread_id=?", (thread_id,))
+        messages = json.loads(self.cursor.fetchone()[0])
+        if message_index < len(messages):
+            return messages[message_index]
+        return None
+
+    def modify_message(self, thread_id: int, message_index: int, new_message_data: Dict[str, Any]):
+        try:
+            serialized_new_message_data = json.dumps(new_message_data)
+        except TypeError as e:
+            print(f"Error serializing new_message_data: {e}")
+            # Handle non-serializable new_message_data appropriately here
+            new_message_data = {k: str(v) for k, v in new_message_data.items()}
+            serialized_new_message_data = json.dumps(new_message_data)
+        
+        self.cursor.execute("SELECT messages FROM ThreadMessages WHERE thread_id=?", (thread_id,))
+        messages = json.loads(self.cursor.fetchone()[0])
+        if message_index < len(messages):
+            messages[message_index] = json.loads(serialized_new_message_data)
+            self.cursor.execute("UPDATE ThreadMessages SET messages=? WHERE thread_id=?", (json.dumps(messages), thread_id))
+            self.conn.commit()
+
+    def remove_message(self, thread_id: int, message_index: int):
+        self.cursor.execute("SELECT messages FROM ThreadMessages WHERE thread_id=?", (thread_id,))
+        messages = json.loads(self.cursor.fetchone()[0])
+        if message_index < len(messages):
+            del messages[message_index]
+            self.cursor.execute("UPDATE ThreadMessages SET messages=? WHERE thread_id=?", (json.dumps(messages), thread_id))
+            self.conn.commit()
+
+    def list_messages(self, thread_id: int) -> List[Dict[str, Any]]:
+        self.cursor.execute("SELECT messages FROM ThreadMessages WHERE thread_id=?", (thread_id,))
+        messages = json.loads(self.cursor.fetchone()[0])
+        return messages
+
+    def run_thread(self, thread_id: int, system_message: Dict[str, Any], model_name: Any, json_mode: bool = False, temperature: int = 0, max_tokens: Optional[Any] = None, tools: Optional[List[Dict[str, Any]]] = None, tool_choice: str = "auto") -> Any:
+        messages = self.list_messages(thread_id)
+        temp_messages = [system_message] + messages
+        response = make_llm_api_call(temp_messages, model_name, json_mode, temperature, max_tokens, tools, tool_choice)
+
+        if tools is None:
+            response_content = response.choices[0].message['content']
+            self.add_message(thread_id, {"role": "assistant", "content": response_content})
+        else:
+            try:
+                response_message = response.choices[0].message
+                tool_calls = response_message.tool_calls
+                print("\nLength of tool calls", len(tool_calls))
```

Comment:
In other places I'm seeing logger, so I guess are we logger logging or print logging?